pkg/controller/appdefinition/checkimageallowed.go

```diff
@@ -0,0 +1,65 @@
+package appdefinition
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	"github.com/acorn-io/acorn/pkg/condition"
+	"github.com/acorn-io/acorn/pkg/imageallowrules"
+	"github.com/acorn-io/baaah/pkg/router"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sirupsen/logrus"
+)
+
+// ImageAllowed is a middleware that checks if the image is allowed by the image allow rules.
+func ImageAllowed(h router.Handler) router.Handler {
+	return router.HandlerFunc(func(req router.Request, resp router.Response) error {
+		appInstance := req.Object.(*v1.AppInstance)
+		for _, cond := range appInstance.Status.Conditions {
+			if cond.Type == v1.AppInstanceConditionImageAllowed && !cond.Success {
+				return fmt.Errorf("image %s is not allowed by rule %s", appInstance.Spec.Image, cond.Reason)
+			}
+		}
+		return h.Handle(req, resp)
+	})
+}
+
+// CheckImageAllowedHandler is a router handler that checks if the image is allowed by the image allow rules and sets a status field accordingly
+func CheckImageAllowedHandler(transport http.RoundTripper) router.HandlerFunc {
+	return func(req router.Request, resp router.Response) error {
+
+		appInstance := req.Object.(*v1.AppInstance)
+		cond := condition.Setter(appInstance, resp, v1.AppInstanceConditionImageAllowed)
+
+		targetImage, unknownReason := determineTargetImage(appInstance)
+		if targetImage == "" {
+			if appInstance.Status.AppImage.Name != "" {
+				targetImage = appInstance.Status.AppImage.Name
+			} else {
+				if unknownReason != "" {
+					cond.Unknown(unknownReason)
+				} else {
+					cond.Error(fmt.Errorf("no image specified"))
+				}
+				return nil
+			}
+		}
+
+		if err := imageallowrules.CheckImageAllowed(req.Ctx, req.Client, appInstance.Namespace, targetImage, remote.WithTransport(transport)); err != nil {
+			if errors.Is(err, &imageallowrules.ErrImageNotAllowed{}) {
+				logrus.Warning(err)
```

Comment:
It seems that if we get an `imageallowrules.ErrImageNotAllowed` then our logic has worked as expected. Meaning that logging a warning here is possibly not warranted?

Reply:
Yeah, it's more useful for Acorn admins that get complaints from users.
But actually I improved the user-facing error messages recently, making this irrelevant as it adds no extra information.